### PR TITLE
Noop locking

### PR DESCRIPTION
**Pull request by furuholm**
This PR implements the noop locking part of #49.

### CODE REVIEW

**Comment by furuholm** [comment]
I added a static assertion as suggested in #49 but I am not sure I did it correctly. First of all the docs says that the following is to be added to lib.rs
```
#[macro_use]
extern crate static_assertions;
```
This lead to a ```unused `#[macro_use]` import``` warning so I didn't include those lines (the macro still worked).

I then verified that `static_assertions::assert_not_impl_all!(Context: Send, Sync);` worked by adding `Send` and `Sync` implementations for Context. As expected this caused a compilation error but the error message not very clarifying.

```
type annotations needed for `fn() {<context::Context as context::_::{{closure}}#0::AmbiguousIfImpl<_>>::some_item}
```

This may very well be correct, but I wasn't sure so it would be great if you could have an extra look at this.

**Comment by furuholm** [comment]
I think the build failure is caused by https://github.com/rust-lang/cargo/pull/7507. I will try to fix this in a separate PR in the coming days.

File: libsignal-protocol/src/context.rs [hunk 1]
```diff
@@ -309,6 +308,14 @@ impl Default for Context {
     }
 }
 
+// Assert that Context does not implement [`Send`] and [`Sync`] to guarantee
+// that each [`Context`] instance is only used within a single thread. This allows
+// locking to be omitted (see lock_function below).
+//
+// See https://github.com/Michael-F-Bryan/libsignal-protocol-rs/issues/49
+// for details.
+static_assertions::assert_not_impl_all!(Context: Send, Sync);
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/src/context.rs]: Could this line be moved directly above `lock_function()`? That way our assertion that `Context: !Send + !Sync` is right next to the comment that says locking isn't required as long as `Context` doesn't implement `Send` and `Sync`.

**Comment by furuholm** [comment]
The build failure is fixed by #52. I would like to have that merged before this change to not fail the build on master.

File: libsignal-protocol/src/context.rs (hunk 1 above)
Comment (furuholm) [reply libsignal-protocol/src/context.rs]: Agree. This looks better.

Fixed in eddb1cd1d

**Comment by Michael-F-Bryan** [comment]
@furuholm I've merged #52. If you do a `rebase` we should be good to merge.

**Comment by furuholm** [comment]
@Michael-F-Bryan rebase done...

**Comment by furuholm** [comment]
Build still fails for `Rust: 1.34.0` though. Seems that the `static_assertions` lib does not compile.

**Comment by Michael-F-Bryan** [comment]
Looking at the [`.travis.yml`](https://github.com/nvzqz/static-assertions-rs/blob/9ee3b1ec857063c8289b2f7014c79d50da09b328/.travis.yml#L4) for `nvzqz/static-assertions-rs` I think we need to bump the minimum `rustc` version to `1.37.0`. 

Would you be able to update [this line](https://github.com/Michael-F-Bryan/libsignal-protocol-rs/blob/f7947d76fa88bd016a0190adc7bd8e28d21e301e/.travis.yml#L41) with the new version and also add a comment saying it's needed for `static_assertions`?

**Pull request merged**